src/compiler/moduleNameResolver.ts

```diff
@@ -1536,11 +1564,114 @@ namespace ts {
         return withPackageId(packageInfo, loadNodeModuleFromDirectoryWorker(extensions, candidate, onlyRecordFailures, state, packageJsonContent, versionPaths));
     }
 
+    /* @internal */
+    export function getEntrypointsFromPackageJsonInfo(
+        packageJsonInfo: PackageJsonInfo,
+        options: CompilerOptions,
+        host: ModuleResolutionHost,
+        cache: ModuleResolutionCache | undefined,
+        resolveJs?: boolean,
+    ): string[] | false {
+        if (!resolveJs && packageJsonInfo.resolvedEntrypoints !== undefined) {
+            // Cached value excludes resolutions to JS files - those could be
+            // cached separately, but they're used rarely.
+            return packageJsonInfo.resolvedEntrypoints;
+        }
+
+        let entrypoints: string[] | undefined;
+        const extensions = resolveJs ? Extensions.JavaScript : Extensions.TypeScript;
+        const features = getDefaultNodeResolutionFeatures(options);
+        const requireState: ModuleResolutionState = {
+            compilerOptions: options,
+            host,
+            traceEnabled: isTraceEnabled(options, host),
+            failedLookupLocations: [],
+            packageJsonInfoCache: cache?.getPackageJsonInfoCache(),
+            conditions: ["node", "require", "types"],
+            features,
+        };
+        const requireResolution = loadNodeModuleFromDirectoryWorker(
+            extensions,
+            packageJsonInfo.packageDirectory,
+            /*onlyRecordFailures*/ false,
+            requireState,
+            packageJsonInfo.packageJsonContent,
+            packageJsonInfo.versionPaths);
+        entrypoints = append(entrypoints, requireResolution?.path);
+
+        if (features & NodeResolutionFeatures.Exports && packageJsonInfo.packageJsonContent.exports) {
+            const exportState = { ...requireState, failedLookupLocations: [], conditions: ["node", "import", "require", "types"] };

```

Comment:
Setting `import` and `require` at the same time seems... Odd. Definitely doesn't match any node runtime behavior. What's the intention?

Reply:
I just want to pull in all the files that a user might auto-import, which includes exports reachable under either mode, since this single project will be queried in both CJS- and ESM-importing locations.

Reply:
This'll only fetch one of them - whichever is higher in the exports object. You'd need to run the call twice, once with each set of conditions.

Reply:
Doh, good catch. These conditions are only consumed in new functions that I wrote, but that is indeed how I wrote them. I could modify those functions instead, but calling it twice is probably the easiest way to ensure we pull in exactly what might be resolved to in either importing mode.